libraries/tlv-account-resolution/src/state.rs

```diff
@@ -158,65 +172,92 @@ impl ExtraAccountMetas {
+
+        for extra_meta in extra_account_metas.data().iter() {
+            let mut account_meta = match extra_meta.discriminator {
+                0 => AccountMeta::try_from(extra_meta)?,
+                1 => {
+                    let seeds = Seed::unpack_array(&extra_meta.address_config)?;
+                    AccountMeta {
+                        pubkey: Self::resolve_pda(
+                            &instruction.accounts,
+                            &instruction.data,
+                            &instruction.program_id,
+                            &seeds,
+                        )?,
+                        is_signer: extra_meta.is_signer.into(),
+                        is_writable: extra_meta.is_writable.into(),
+                    }
+                }
+                _ => return Err(ProgramError::InvalidAccountData),
+            };
+            Self::de_escalate_account_meta(&mut account_meta, &instruction.accounts);
+            instruction.accounts.push(account_meta);
+        }
+
+        Ok(())
     }
 
-    /// Add the additional account metas and account infos for a CPI, while
-    /// de-escalating repeated accounts.
-    ///
-    /// If an added account already exists in the instruction with lower
-    /// privileges, match it to the existing account. This prevents a lower
-    /// program from gaining unexpected privileges.
+    /// Add the additional account metas and account infos for a CPI
     pub fn add_to_cpi_instruction<'a, T: SplDiscriminate>(
         cpi_instruction: &mut Instruction,
         cpi_account_infos: &mut Vec<AccountInfo<'a>>,
         data: &[u8],
         account_infos: &[AccountInfo<'a>],
     ) -> Result<(), ProgramError> {
-        let state = TlvStateBorrowed::unpack(data)?;
-        let bytes = state.get_first_bytes::<T>()?;
-        let extra_account_metas = PodSlice::<PodAccountMeta>::unpack(bytes)?;
+        let initial_instruction_metas = cpi_instruction.accounts.clone();
 
-        let initial_cpi_instruction_length = cpi_instruction.accounts.len();
+        Self::add_to_instruction::<T>(cpi_instruction, data)?;
 
-        for mut account_meta in extra_account_metas.data().iter().map(AccountMeta::from) {
+        for account_meta in cpi_instruction.accounts.iter().filter(|&x| {
+            !initial_instruction_metas
+                .iter()
+                .any(|y| y.pubkey == x.pubkey)
+        }) {
             let account_info = account_infos
                 .iter()
                 .find(|&x| *x.key == account_meta.pubkey)
                 .ok_or(AccountResolutionError::IncorrectAccount)?
                 .clone();
-            Self::de_escalate_account_meta(
-                &mut account_meta,
-                &cpi_instruction.accounts,
-                initial_cpi_instruction_length,
-            );
-            cpi_account_infos.push(account_info);
-            cpi_instruction.accounts.push(account_meta);
+            cpi_account_infos.push(account_info.clone());
```

Comment:
Is the second clone needed? I'm surprised that clippy didn't say anything
```suggestion
            cpi_account_infos.push(account_info);
```